### ISSUE TITLE
Windows: Ideas to make a faster DLL dependency caching system

### ISSUE THREAD

**Issue by brupelo** [Nuitka/Nuitka#85]
Right now the caching system on windows is far away of being optimal, to expose one of the multiple problems let's run the next test:

**Experiment:**

 - Given 2 equal projects, [test1](https://github.com/brupelo/nuitka_tests/tree/master/test1) and [test1_clone](https://github.com/brupelo/nuitka_tests/tree/master/test1_clone)
 - Given 2 virtualenvs with the same requirements.txt, [venv](https://github.com/brupelo/nuitka_tests/blob/master/venv.bat) & [venv_clone](https://github.com/brupelo/nuitka_tests/blob/master/venv_clone.bat)

**Steps (running build.bat on each step):**

 - Initial state: Cache not created
 - (venv) test1 (cache miss): Execution time: 377.859 s
 - (venv) test1 (cache hit): Execution time: 10.121 s  (37x faster building time)
 - (venv) test1_clone (cache hit): Execution time: 10.260 s
 - (venv_clone) test1 (cache miss): Execution time: 390.493 s
 - (venv_clone) test1 (cache hit): Execution time: 10.194 s
 - (venv_clone) test1_clone (cache hit): Execution time: 10.932 s

**Problem:**

As you can see, when using the same virtualenv/system_wide python to build several projects, that can or can't be structurally equal (in terms of dependencies) the caching system works relatively well, as right now is creating cache keys at a level of path (venv/system_wide) the cache will be hit with a high probability.

The problem appears when using different virtualenvs for the same projects, there will be more cache miss than necessary. As you can see in the above sequence there are 2 cache misses when there should be only 1! And this wouldn't be a problem if depends was fast enough but right now one cache miss is quite expensive. For instance, one cache miss for the trivial test1 costs 380s, that's just crazy...

**Solution:**

Consider file checksums to create cache keys, ie:

    def md5(fname):
        hash_md5 = hashlib.md5()
        with open(fname, "rb") as f:
            for chunk in iter(lambda: f.read(4096), b""):
                hash_md5.update(chunk)
        return hash_md5.hexdigest()

as well as normalized paths, for instance, let's say the same dll is living in different virtualenvs (venv1, venv2, ..., venv_n) one possible cache key could be something like this:

        2fac067588e301242f8c1ef2b1255594: ['path/to/venv1/../fancydll.dll', 'path/to/venv2/../fancydll.dll', ... , 'path/to/venv_n/../fancydll.dll']

That entry is saying that fancydll.dll with checksum 2fac067588e301242f8c1ef2b1255594 could exist in different places or not. Why not? Well, you're creating/destroying virtualenvs over and over, in case it doesn't exist in any path of the list it means you're using a non-cache virtualenv/system-wide python location.

Anyway, the explained above is intended to show you what's one of the big problems of the current caching system and one possible idea to think about... If depends.exe wasn't that slower maybe you wouldn't even need a caching system at all but right now let's consider depends.exe is the main bottleneck when building a package.

NS: Already gave you my opinion in another thread about debug options like `disable-dll-dependency-cache`, they're great to use in debug mode but definitely they shouldn't live in production at all, users don't care at all about internals of the software, just give them as little available options as possible, period... ;)  (IMHO)

**Comment by kayhayen** [comment]
Actually what I was thinking, is to detect the wheel something came from, and handle things based on that, process the wheel, adding information for it. I kind of wish there would be a result that could even be distributed with them, in the form of `.pyi` files.

One thing that `depends.exe` gives is full paths to DLLs. So the Qt DLL paths will point inside the venv, and be valid for only as long as it exists. Making those PATHs relative will be extra work. Right now the raw result is cached.

And then, generally, changes to user PATH, should affect compilation. They kind of should mean that the cache is invalid fully. Need I mention that a venv has its own PATH element. Normalizing PATH some, and removing elements of it, where no DLLs live, could be done.

This all seems to be a lot of effort that somebody else would have to do. I would prefer for a solution to aim at producing something we can make "sip" and or "wheel" include for upstream. For imported modules, ".pyi" would already work. For used DLLs, we would have to come up with something for them to include. That might be even more effort.

So from my side, this is not something I want to invest time in unless there were really big donations attached.

An alternative would be, for a given project, to create a "dist" file and make the user keep it, and provide it. Maybe giving warnings if something changed. It would be system specific, or made to use more abstract information, which is when we enter the business of location DLLs according to Windows search path rules. Also a lot of work.

So from my side, this is also not something I want to invest my time in right now.

**Comment by brupelo** [comment]
> Actually what I was thinking, is to detect the wheel something came from, and handle things based on that, process the wheel, adding information for it. I kind of wish there would be a result that could even be distributed with them, in the form of .pyi files.

Talking about this one, few days ago I'd answered ([bad answer btw](https://stackoverflow.com/a/50216621/3809375), as the good answers were already taken) a question asking about wheels, this package looks [handy](https://stackoverflow.com/a/50214191/3809375).

Anyway, this issue is easy and interesting to me enough. Eventually (few months or so) I could invest my spare time on creating a little weekend hobby project about it, I'd like to code my custom own pe parser, I'm quite interesting to learn more about the inner details of pe files.

One of the things I'd use that parser to is to write my own exe compressor and that parser could also be used to write a dll dep analizer where you'd got full control. I'd try to make it as fast as possible so let me be speculate a bit here and caching probably wouldn't be needed. Not sure how hard this would be but believe me... After seeing the crazy high times of depends.exe with a cold cache, I believe than spawning many times depends.exe is not good, with a custom dependency analizer you won't only have the overhead of spawning an external exe over and over but you also will be able input all the gathered interested data at once in the same nuitka process. An advice I can give from experience that it's always better spawning 1 process to handle n tasks than n processes to handle n tasks.

In any case, this will have to wait, I've spent quite a lot of time these days testing nuitka but unfortunately I've got many other personal projects (work and hobby ones) to handle so my time is kind of limited. The good thing about testing nuitka all these days has allowed me to have at least a wider perspective of what's the current status of the project right now, I can see there is quite a lot of pending job to do.... Hopefully it'll get more code contributors this year and becoming more popular, I think even if the project isn't yet mature enough to be used on production is mature enough to become more popular along the python community.

Yours,
Bruno

**Comment by brupelo** [comment]
Ok, I couldn't wait to at least make a fast test lol :'D. 

I was really curious about this one so instead writing my own pe-parser from scratch I've made a fast test using [pefile](https://github.com/erocarrera/pefile) library, after seeing the ammount of contributors, number of commits and projects used by that library I've realized if I wanted to learn more about pe file internals better learn from some well tested thing like that instead reinveinting the wheel once again, although tbh I do really love reinventing the wheel and I make it all the time just for the sake of fun/learning :'D

Anyway, i've used that library to analize deps from all pyd/dll files living in a virtualenv with pyqt/pyopengl installed and also c:\windows\system32. The results were quite awesome ones, take a [look](https://dl.dropboxusercontent.com/s/t0xat0zvq3ehgzl/2018-05-12_13-19-38.txt). First column is the index of the dll, second how long took to gather the deps of that particular file and the third the size of that dll. I won't make any gaussian to display it but the more common case is times living in the order of few ms.

Summing up, now we know than using depends.exe is a really bad choice for windows :) . Although if you ask me I can understand why it's taking so long to depends.exe to analize deps from the command line... So yeah, good time to throw depends.exe to the bin and never look back again (is that proper English) ;)

Best,
B

**Comment by kayhayen** [comment]
@brupelo Seeing pyinstaller being on the list of things that uses pefile, it's probably more viable than I think. If you manage to replace dependency walker, and replace it with what might be pure Python (is it?) that will be great.

**Comment by brupelo** [comment]
@kayhayen Sure, I think this ticket is quite easy actually and I could give it a shot. Eventually (can't promise when) I'll find time to give it a shot to it as I'm quite interested on boost up nuitka myself and having more control about this dependency stuff ;) . So forget about this one, I'll handle it.

> be pure Python (is it?) 

Yup, as an additional note, I don't think we'd benefit here of writing/wrapping an asm/c/c++ pe parser into python... as python is relatively fast parsing files like pe ones.

**Comment by kayhayen** [comment]
This is now so fast, it needs no caching anymore, so this has become obsolete, plus the caching code is actually still there, just disabled.
